### PR TITLE
[AIC-py][eval] small refactor to ease defining metrics, add tests

### PR DESCRIPTION
**Pull request by jonathanlastmileai**
[AIC-py][eval] small refactor to ease defining metrics, add tests

- Add decorators for metric creation and reimplement some existing ones
- Add a unit test
- I'm also bumping to 1.1.15 so we can publish from trunk tomorrow and unblock CircleCI.

Test plan:

Existing and new unit tests.

### CODE REVIEW

File: python/src/aiconfig/eval/metrics.py [hunk 1]
```diff
@@ -29,22 +29,65 @@ async def __call__(self, datum: common.T_Evaluable) -> common.T_MetricValue:
...
+@core_utils.parametrized
+def metric(
+    parametrized_evaluation_fn: Callable[Concatenate[common.T_Evaluable, PS], common.T_MetricValue],
+    name: str | None = None,
+    description: str | None = None,
+    best_value: common.T_MetricValue | None = None,
+    worst_value: common.T_MetricValue | None = None,
+) -> Callable[PS, Metric[common.T_Evaluable, common.T_MetricValue]]:
+    name_ = name or parametrized_evaluation_fn.__name__
+    description_ = description or name_
```
Comment (rholinshead) [inline python/src/aiconfig/eval/metrics.py:44]: nit, would probably be good to differentiate the description text from name. Maybe default to f"{name} evaluation metric"?

File: python/src/aiconfig/eval/api/__init__.py [hunk 2]
```diff
@@ -13,15 +13,12 @@
...
-    "Metric",
     "common",
     "metrics",
-    "brevity",
-    "substring_match",
+    "TestSuiteWithInputsSettings",
     "run_test_suite_with_inputs",
     "run_test_suite_outputs_only",
-    "TestSuiteWithInputsSettings",
+
```
Comment (rossdanlm) [inline python/src/aiconfig/eval/api/__init__.py:23]: nit: alphabetical ordering + remove newline

File: python/src/aiconfig/eval/metrics.py [hunk 3]
```diff
@@ -29,22 +29,65 @@ async def __call__(self, datum: common.T_Evaluable) -> common.T_MetricValue:
...
+@core_utils.parametrized
+def metric_async(
+    parametrized_evaluation_fn: Callable[Concatenate[common.T_Evaluable, PS], Awaitable[common.T_MetricValue]],
+    name: str | None = None,
+    description: str | None = None,
+    best_value: common.T_MetricValue | None = None,
+    worst_value: common.T_MetricValue | None = None,
+) -> Callable[PS, Metric[common.T_Evaluable, common.T_MetricValue]]:
+    name_ = name or parametrized_evaluation_fn.__name__
+    description_ = description or name_
```
Comment (rholinshead) [inline python/src/aiconfig/eval/metrics.py:73]: same here

File: python/src/aiconfig/eval/metrics.py [hunk 4]
```diff
@@ -29,22 +29,65 @@ async def __call__(self, datum: common.T_Evaluable) -> common.T_MetricValue:
...
-    output_datum: str,
-    substring: str,
-    #
-    case_sensitive: bool,
-) -> bool:
-    if case_sensitive:
-        return substring in output_datum
-    else:
-        return substring.lower() in output_datum.lower()
+PS = ParamSpec("PS")
```
Comment (rossdanlm) [inline python/src/aiconfig/eval/metrics.py:32]: nit: `T_ParamSpec`

File: python/tests/test_eval.py [hunk 5]
```diff
@@ -188,6 +190,79 @@ async def test_run_test_suite_with_inputs(data: st.DataObject):
             assert False, f"expected Ok, got Err({e})"
 
 
+@hypothesis.given(st.data())
+@pytest.mark.asyncio
+async def test_run_test_suite_with_inputs_general_params(data: st.DataObject):
+    """In test_run_test_suite_outputs_only, we test the user-facing function (e2e)
+    In this case that's harder because run_test_suite_with_inputs takes
```
Comment (rholinshead) [inline python/tests/test_eval.py:197]: Couldn't you mock fs load of the path to return mock json?

**Review by rholinshead — APPROVED** [review]
Just minor things

File: python/src/aiconfig/eval/lib.py [hunk 6]
```diff
@@ -162,7 +162,7 @@ def _extract_value(input_text_datum: TextBasedInputDatum | None) -> str | None:
                     case str(input_text):
                         return input_text
                     case frozendict():
-                        return json.dumps(input_text_datum.value)
+                        return json.dumps(input_text_datum.value, sort_keys=True)
```
Comment (rossdanlm) [inline python/src/aiconfig/eval/lib.py:165]: Nit: needed for serialization of dict --> text for testing, and also have it be deterministitc in canonical form

File: python/src/aiconfig/eval/metrics.py [hunk 7]
```diff
@@ -261,37 +304,37 @@ def make_openai_structured_llm_metric(
...
-            name="substring_match",
-            description="True (pass) if contains given substring",
-            best_value=True,
-            worst_value=False,
-            extra_metadata=dict(substring=substring, case_sensitive=case_sensitive),
-        ),
-    )
 
+@metric(
+    #
```
Comment (rossdanlm) [inline python/src/aiconfig/eval/metrics.py:311]: nit: remove

File: python/tests/mocks.py [hunk 8]
```diff
@@ -1,28 +1,44 @@
...
+
+            Real method: https://github.com/lastmile-ai/aiconfig/blob/a4376d1f951e19776633d397a3cda7fa85506eef/python/src/aiconfig/Config.py#L277
+            """
+            params_ = params or {}
+            return await mock_run_text_to_text_impl(prompt_name, params_)
+
+    return _MockAIConfigRuntime()
+
+
+MockAIConfigRuntime = lambda: make_mock_aiconfig_runtime()
```
Comment (rossdanlm) [inline python/tests/mocks.py:44]: nit: remove lambda

File: python/tests/test_eval.py [hunk 9]
```diff
@@ -246,11 +321,12 @@ async def test_exception_metric(caplog: pytest.LogCaptureFixture):
     user_test_suite_outputs_only = list(
         itertools.product(
             ["Hundred Acre Wood", ""],
-            [metrics.brevity],
+            [brevity],
         )
     )
     with caplog.at_level(logging.ERROR):
         df = await run_test_suite_outputs_only(user_test_suite_outputs_only)
+        print(df[["metric_name"]])
```
Comment (rossdanlm) [inline python/tests/test_eval.py:329]: nit: remove

File: python/tests/test_eval.py [hunk 10]
```diff
@@ -188,6 +190,79 @@ async def test_run_test_suite_with_inputs(data: st.DataObject):
...
+                df[["input", "metric_id"]].itertuples(index=False, name=None)  # type: ignore[no-untyped-call]
+            )
+
+            assert input_pairs == result_pairs, f"fail: {input_pairs=}, {result_pairs=}"
+
+            df_brevity = df[df["metric_name"] == "brevity"]  # type: ignore
+            assert (df_brevity["aiconfig_output"].apply(len) == df_brevity["value"]).all()  # type: ignore
+
+            df_substring = df[df["metric_name"] == "substring_match"]  # type: ignore
+            assert (df_substring["value"].apply(lambda x: x in {False, True})).all()  # type: ignore
```
Comment (rossdanlm) [inline python/tests/test_eval.py:260]: TODO (rossdan): Implement checking on substring for actual value, not just type check that this is boolean

**Review by rossdanlm — APPROVED** [review]
3d349795-639d-467e-b4a8-862a8e6493db

Nice!

File: python/tests/test_eval.py (hunk 5 above)
Comment (jonathanlastmileai) [reply python/tests/test_eval.py:197]: Yeah I guess I could. Adding here

File: python/src/aiconfig/eval/metrics.py (hunk 7 above)
Comment (jonathanlastmileai) [reply python/src/aiconfig/eval/metrics.py:311]: I'm going to just make the line length shorter. cc @Ankush-lastmile

**Pull request merged**